File: backend/dataset/numIslands/numislands_7.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid: return 0
        m, n = len(grid), len(grid[0])
        ans = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    q = [(i, j)]
                    while q:
                        x, y = q.pop()
                        grid[x][y] = '2'
                        for dx, dy in (0, 1), (0, -1), (1, 0), (-1, 0):
                            xx, yy = x+dx, y+dy
                            if 0 <= xx < m and 0 <= yy < n and grid[xx][yy] == '1':
                                q.append((xx, yy))
                    ans += 1            
        return ans

File: backend/CDataset/numIslands/numIslands_7.cpp
class Solution {
public:
    int count=0;
    int numIslands(vector<vector<char>>& grid) {
        int row=grid.size();
        int col=grid[0].size();
        for (int i=0; i<row; i++) {
            for (int j=0; j<col; j++) {
                if (grid[i][j]=='1') count++, dfs (grid, i, j) ;
            }
        }
        return count;
    }
    void dfs (vector<vector<char>> &grid, int i, int j) {
        if (i<0 || j<0 || i>=grid.size() || j>=grid[0].size() || grid[i][j]!='1') return;
        grid[i][j]='0';
        vector <vector<int>> v={{0,1}, {1,0}, {-1,0}, {0,-1}};
        for (int x=0; x<4; x++) {
            dfs (grid, i+v[x][0], j+v[x][1]);
        }
    }
};
